zip-0315.rst

```diff
@@ -1,9 +1,629 @@
+indistinguishable, then we can't auto-shield when the transparent balance reaches
+some threshold (otherwise opportunistic would either never be used, or would be
+identifiable when it uses the balance below the threshold).
+
+Instead, a proposition: we define a distribution of "time since last payment to the
+address" from which we sample the time at which the auto-shielding transaction will
+be created. This distribution is weighted by the balance in the address, so as more
+funds accrue, the auto-shielding transaction is more likely to be created.
+
+- It ensures that all funds will eventually be auto-shielded, while preventing
+  fee-dusting attacks (where dust is sent in order to repeatedly consume fees from
+  the wallet), as the auto-shielding transaction is not directly triggered by payment
+  receipt.
+
+- If the user makes a shielding transaction in the meantime, we opportunistically
+  shield, without it being clearly not an auto-shielding transaction.
+
+- If a wallet is offline for a long time, then it would likely auto-shield as soon as
+  it finishes syncing. This maybe isn't enough to reveal that the wallet came online,
+  except that it _might_ result in auto-shielding transactions for multiple
+  transparent addresses being created at the same time. So we might want to
+  special-case this?
+
+Properties we want from auto-shielding:
+
+- Auto-shielding transactions MUST NOT shield from multiple transparent receivers in
+  the same transaction.
+  - Doing so would trivially link diversified UAs containing transparent receivers.
+
+Properties we want from auto-migration:
+
+- Receipt of a shielded payment MUST NOT trigger any on-chain behaviour (as that
+  reveals transaction linkability).
+
+Both auto-shielding and auto-migration are time-triggered actions, not
+receipt-triggered actions. An auto-shielding or auto-migration transaction MUST NOT
+be created as a direct result of a payment being received.
+
+Both of these are opportunistic: if the user's wallet is making a transaction in
+which one of these actions would occur anyway, then the wallet takes the opportunity
+to migrate as much as it would do if it were generating an autoshielding transaction.
+This both saves on a transaction, and removes the need for any kind of transparent
+change address within UAs.
+
+TODO: what pool should change go to?
+
+* Proposal: the most recent pool already involved in the transaction.
+
+Wallet Recovery
+---------------
+
+In the case where we are recovering a wallet from a backed-up mnemonic phrase,
+and not from a wallet.dat, we don't have enough information to figure out what
+receiver types the user originally used when deriving each UA under an account.
+We have a similar issue if someone exports a UFVK, derives an address from it,
+and has a payment sent to the address: zcashd will detect the payment, but has
+no way to figure out what address it should display in the UI. A wallet could
```

Comment:
If the problem is what address to show as the receiving address in an incoming transaction, my first question is why would the user care to see any of the various forms of their own address. Isn't it enough to know it's coming in to their wallet?
But I actually have a doc that describes how this could be useful, but without having to deal with this problem. But it relies on off-chain data, and your problem statement begins with supposing wallet.dat isn't available. Without the off-chain data, my spec would be limited to giving the timestamp at which the sender first received the address to use, based on time-based diversifiers being given every time a receiving address is handed out. My data file will tie that diversifier index to the contact in the address book that was given the receiving address for absolute correlation. But without that file, we still could say "you received payment from someone using an address you generated on [some date/time]".

Reply:
The main reason would be for when a user provides distinct diversified addresses to different third parties; in this case, the address encodes some information about who the payment might have come from (or at least who might have had access to that particular address.)

There should definitely be a section here that discusses UX around diversified addresses, if there isn't already.

Reply:
Right. I agree that can be useful. But as the user isn't likely to keep offline or hand-written copies of each of these diversified addresses that they hand out, they'll be relying on the wallet software to help them correlate incoming payments with those addresses or rather the people or places that observed those addresses. So whether the wallet can show the original unified address or just one of the receivers doesn't impact the wallet's ability to draw those correlations, IMO.